src/NuGet.Core/NuGet.Commands/RestoreCommand/CompatibilityChecker.cs

```diff
@@ -301,16 +316,94 @@ private static bool IsProjectCompatible(Library library)
             }
         }
 
-        private bool IsCompatible(CompatibilityData compatibilityData)
+        private bool IsPackageCompatible(CompatibilityData compatibilityData)
         {
             // A package is compatible if it has...
             return
                 HasCompatibleAssets(compatibilityData.TargetLibrary) ||
                 !compatibilityData.Files.Any(p =>
                     p.StartsWith("ref/", StringComparison.OrdinalIgnoreCase)
-                    || p.StartsWith("lib/", StringComparison.OrdinalIgnoreCase));                       // No assemblies at all (for any TxM)
+                    || p.StartsWith("lib/", StringComparison.OrdinalIgnoreCase)); // No assemblies at all (for any TxM)
+        }
+
+        private static HashSet<FrameworkRuntimePair> GetAvailableFrameworkRuntimePairs(CompatibilityData compatibilityData, RestoreTargetGraph graph)
+        {
+            var available = new HashSet<FrameworkRuntimePair>();
+
+            var contentItems = new ContentItemCollection();
+            contentItems.Load(compatibilityData.Files);
+
+
+            if (compatibilityData.TargetLibrary.PackageType.Contains(PackageType.DotnetTool))
+            {
+                foreach (var group in contentItems.FindItemGroups(graph.Conventions.Patterns.ToolsAssemblies))
+                {
+                    group.Properties.TryGetValue(ManagedCodeConventions.PropertyNames.RuntimeIdentifier, out var ridObj);
+                    group.Properties.TryGetValue(ManagedCodeConventions.PropertyNames.TargetFrameworkMoniker, out var tfmObj);
+
+                    var tfm = tfmObj as NuGetFramework;
+                    var rid = ridObj as string;
+                    if (tfm?.IsSpecificFramework == true)
+                    {
+                        available.Add(new FrameworkRuntimePair(tfm,rid));
+                    }
+                }
+            }
+
+            return available;
         }
 
+        private async Task VerifyDotnetToolCompatibilityChecks(CompatibilityData compatibilityData, GraphItem<RemoteResolveResult> node, RestoreTargetGraph graph, List<CompatibilityIssue> issues)
+        {
+            var containsDotnetToolPackageType = compatibilityData.TargetLibrary.PackageType.Contains(PackageType.DotnetTool);
+
+            if (compatibilityData.TargetLibrary.PackageType.Count != 1 && containsDotnetToolPackageType)
+            {
+                var issue = CompatibilityIssue.ToolsPackageWithExtraPackageTypes(
+                    new PackageIdentity(node.Key.Name, node.Key.Version));
+
+                issues.Add(issue);
+                await _log.LogAsync(GetErrorMessage(NuGetLogCode.NU1204, issue, graph));
+            }
+
+            if (containsDotnetToolPackageType &&
+                    !(HasCompatibleToolsAssets(compatibilityData.TargetLibrary) || !compatibilityData.Files.Any(p => p.StartsWith("tools/", StringComparison.OrdinalIgnoreCase))))
```

Comment:
since you do a check for files fairly often, can we abstract out functions that would tell us if there are any lib files or tool files like this: 

compatibilityData.HasLibFiles
compatibilityData.HasToolsFiles etc etc.

Reply:
HasCompatibleToolsAssetts is a static used once in the compat checker, and once in the lock file builder. 
It can't be bound to the CompatibilityData model.